File: 278. First Bad Version.cpp
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    // binary search, O(logN)time, O(1)space, 40.01%, ok
    int firstBadVersion(int n) {
        int head = 0, tail = n, ret = 0;
        while(head <= tail){
            int mid = head + (tail - head) / 2;
            if(isBadVersion(mid)){
                ret = mid;
                tail = mid - 1;
            }
            else head = mid + 1;
        }
        return ret;
    }
};

File: 278. First Bad Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    # binary search: O(logN) time, O(1) space
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        low, high = 1, n
        while low + 1 < high:
            mid = low + (high - low) / 2
            if isBadVersion(mid):
                high = mid
            else:
                low = mid
        if isBadVersion(low):
            return low
        return high
        